lib/unit_tests.py

```diff
@@ -0,0 +1,185 @@
+   UnitTest with info from the rule directory (rule, test, is_pass)
+
+ - fill_in_metadata() reads the .sh contents and completes UnitTest
+   based on metadata contained in the .sh file
+"""
+
+import re
+import io
+import contextlib
+import collections
+from pathlib import Path
+
+
+UnitTest = collections.namedtuple(
+    'UnitTest',
+    [
+        # str: rule name
+        'rule',
+        # str: test name without .pass.sh or .fail.sh
+        'test',
+        # bool: True if .pass.sh, False if .fail.sh
+        'is_pass',
+        # tuple of str or None: RPM package names
+        'packages',
+        # tuple of str or None: profiles the test should be run within
+        'profiles',
+        # dict or None: variables that should be changed in datastream/playbook
+        'variables',
+        # str or None: which remediation types should be tested (none, bash, ansible, ..)
+        'remediation',
+        # str or None: check engine type
+        'check',
+    ],
+    # last 5 fields are optional
+    defaults=[None, None, None, None, None],
+)
+
+
+# a=b,c=d,e=f  --> {'a': 'b', 'c': 'd', 'e': 'f'}
+# a=b,c        --> {'a': 'b,c'}
+# a=b , c      --> {'a': 'b , c'}           # preserved space
+# a=b, c ,d=e  --> {'a': 'b, c', 'd': 'e'}
+# a=b, c=d     --> {'a': 'b', 'c': 'd'}
+# a=b,c= d     --> {'a': 'b', 'c': 'd'}
+# a=b,c =d     --> {'a': 'b', 'c': 'd'}
+# a=b, c = d   --> {'a': 'b', 'c': 'd'}
+# a=b , c = d  --> {'a': 'b', 'c': 'd'}
+# a=,b=c       --> {'a': '', 'b': 'c'}
+# a=b,=c       --> error key empty
+# a            --> error missing key
+def parse_variables(line):
+    variables = {}
+    key = None
+    for part in line.split(','):
+        if '=' in part:
+            key, _, value = part.partition('=')
+            key = key.strip()
+            value = value.lstrip()
+            if not key:
+                raise ValueError(f"key empty: {part}")
+            variables[key] = value
+        else:
+            if key is None:
+                raise ValueError(f"no key=value pair given: {part}")
+            variables[key] += f',{part}'
+    # trim any trailing spaces in values
+    # (we need to do it here because we might have added space-containing
+    #  equals-less extra values to a previous key via the 'else' branch)
+    for key, value in variables.items():
+        variables[key] = value.rstrip()
+    return variables
+
+
+def fill_in_metadata(unit_test, test_file):
+    """
+    Given an (incomplete) UnitTest tuple as 'unit_test', fill in values
+    from unit test file metadata, returning the complete UnitTest.
+
+    'test_file' may be a file path or a file-like object.
+    """
+    packages = None
+    profiles = None
+    variables = None
+    remediation = None
+    check = None
+
+    with contextlib.ExitStack() as stack:
+        if isinstance(test_file, io.IOBase):
+            f = test_file
+        else:
+            f = stack.enter_context(open(test_file))
+
+        for line in f:
```

Comment:
Note for myself: While not as offensive as `i` in a huge `for` loop, having a single-letter `f` here is probably not great. Also, it's not necessary (after undo-ing metadata-on-top feature), we can just pass file path and get rid of the `ExitStack` logic, because we don't need to call `fill_in_metadata()` with a file stream.